### PR TITLE
Add io.github.qcanvas.QCanvasApp

### PR DESCRIPTION
**Pull request by QCanvas**
### Please confirm your submission meets all the criteria

- [x] Please describe the application briefly.
  QCanvas is a desktop client for canvas LMS which lets you download course content and view it offline.

- [x] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read the and followed all the [Submission and licence requirements][reqs].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor of the project.

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements#control-over-domain
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

### CODE REVIEW

**Comment by QCanvas** [comment]
It's x86-64 only because I had to use https://github.com/flatpak/flatpak-builder-tools/pull/331, otherwise the pip generator tried to build the packages and then failed (probably due to https://github.com/flatpak/flatpak-builder-tools/issues/380)

**Comment by QCanvas** [comment]
bot, build io.github.qcanvas.QCanvasApp

**Comment by QCanvas** [comment]
bot, stop

**Comment by QCanvas** [comment]
bot, help

**Comment by QCanvas** [comment]
bot, abort

**Comment by QCanvas** [comment]
bot, build io.github.qcanvas.QCanvasApp

**Comment by QCanvas** [comment]
Looks like it works

File: io.github.qcanvas.QCanvasApp.desktop
Comment (hfiguiere) [inline io.github.qcanvas.QCanvasApp.desktop:1]: this should be managed upstream

File: io.github.qcanvas.QCanvasApp.metainfo.xml
Comment (hfiguiere) [inline io.github.qcanvas.QCanvasApp.metainfo.xml:1]: this should be managed upstream.

File: flathub.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
+  "only-arches": [
+    "x86_64"
```
Comment (hfiguiere) [inline flathub.json:3]: this is disappointing since this is in Python.

File: io.github.qcanvas.QCanvasApp.svg
Comment (hfiguiere) [inline io.github.qcanvas.QCanvasApp.svg:1]: icon should be upstream

File: python3-qcanvas.json [hunk 2]
```diff
@@ -0,0 +1,244 @@
+{
+    "name": "python3-qcanvas",
+    "buildsystem": "simple",
+    "build-commands": [
+        "pip3 install --verbose --exists-action=i --no-index --find-links=\"file://${PWD}\" --prefix=${FLATPAK_DEST} \"qcanvas\" --no-build-isolation"
+    ],
+    "sources": [
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/b3/e7/ca2993c7682d8629b62630ebf0d1f3bb3d579e667ce8e7ca03a0a0576a2d/Brotli-1.1.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl",
```
Comment (hfiguiere) [inline python3-qcanvas.json:10]: this has a source package and should be used instead.

File: python3-qcanvas.json [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+            "sha256": "e4ff4054e871cfca81bda1cb59561367fa9edba7512806f1ecf81fcc8f125e3f"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/7e/a9/2146d5117ad8a81185331e0809a6b48933c10171f5bac253c6df9fce991c/QtPy-2.4.1-py3-none-any.whl",
+            "sha256": "1c1d8c4fa2c884ae742b069151b0abe15b3f70491f3972698c683b8e38de839b"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/bb/be/4045425132a2fde3422e0d2062e6a3b874e385b98b0442c1d7d29ac152bc/SQLAlchemy-2.0.31-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl",
```
Comment (hfiguiere) [inline python3-qcanvas.json:25]: this has a source package that shouls be used instead.

File: python3-qcanvas.json [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+            "sha256": "3ae3b49a3d5e28a77a0be2b37dbcb89005058959cb2323858c2657c4a8cab474"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/1c/d5/c84e1a17bf61d4df64ca866a1c9a913874b4e9bdc131ec689a0ad013fb36/certifi-2024.7.4-py3-none-any.whl",
+            "sha256": "c198e21b1289c2ab85ee4e67bb4b4ef3ead0892059901a8d5b622f24a1101e90"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/40/26/f35951c45070edc957ba40a5b1db3cf60a9dbb1b350c2d5bef03e01e61de/charset_normalizer-3.3.2-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl",
```
Comment (hfiguiere) [inline python3-qcanvas.json:75]: this too has a source tarball.

File: python3-qcanvas.json [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+            "sha256": "70dda5694a5b194a8441f077aa5fb70cc94e4ec08016117523f013680901ecb7"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/f8/39/e5143e7ec70939d2076c1165ae9d4a3815597019c4d797b7f959cf778600/graphql_core-3.2.3-py3-none-any.whl",
+            "sha256": "5766780452bd5ec8ba133f8bf287dc92713e3868ddd83aee4faab9fc3e303dc3"
+        },
+        {
+            "type": "file",
+            "url": "https://files.pythonhosted.org/packages/c6/1f/12d5a6cc26e8b483c2e7975f9c22e088ac735c0d8dcb8a8f72d31a4e5f04/greenlet-3.0.3-cp311-cp311-manylinux_2_24_x86_64.manylinux_2_28_x86_64.whl",
```
Comment (hfiguiere) [inline python3-qcanvas.json:95]: this one too.

File: utils/metainfo-version-updater.py
Comment (hfiguiere) [inline utils/metainfo-version-updater.py:1]: this belong upstream like the appstream file.

**Comment by hfiguiere** [comment]
> It's x86-64 only because I had to use [flatpak/flatpak-builder-tools#331](https://github.com/flatpak/flatpak-builder-tools/pull/331), otherwise the pip generator tried to build the packages and then failed (probably due to [flatpak/flatpak-builder-tools#380](https://github.com/flatpak/flatpak-builder-tools/issues/380))

Maybe you help and contribute the fixes.

File: io.github.qcanvas.QCanvasApp.metainfo.xml [hunk 6]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="utf-8"?>
+<component type="desktop-application">
+	<id>io.github.qcanvas.QCanvasApp</id>
+	<name>QCanvas</name>
+	<summary>A desktop client for Canvas LMS</summary>
+	<content_rating type="oars-1.0"></content_rating>
```
Comment (hfiguiere) [inline io.github.qcanvas.QCanvasApp.metainfo.xml:6]:
```suggestion
	<content_rating type="oars-1.1"></content_rating>
```

**Comment by QCanvas** [comment]
Using the tarballs instead of the wheels, the build fails with this error, when trying to build `yarl`:
```
ModuleNotFoundError: No module named 'expandvars'
```
Can I somehow include this dependency for the build?

**Comment by QCanvas** [comment]
> > It's x86-64 only because I had to use [flatpak/flatpak-builder-tools#331](https://github.com/flatpak/flatpak-builder-tools/pull/331), otherwise the pip generator tried to build the packages and then failed (probably due to [flatpak/flatpak-builder-tools#380](https://github.com/flatpak/flatpak-builder-tools/issues/380))
> 
> Maybe you help and contribute the fixes.

I don't think I'm capable of that. I'm relatively new to python and that issue seems to be blocked by an upstream pip issue. Additionally, some of the packages I have rust modules and I don't know how to build those

**Comment by QCanvas** [comment]
bot, build io.github.qcanvas.QCanvasApp

File: flathub.json (hunk 1 above)
Comment (hfiguiere) [reply flathub.json:3]: so maybe this can be removed now?

Comment (QCanvas) [reply flathub.json:3]: No, I'm not sure how I will get this to build without using some wheels

**Comment by QCanvas** [comment]
bot, build io.github.qcanvas.QCanvasApp

**Comment by QCanvas** [comment]
bot, build io.github.qcanvas.QCanvasApp

**Comment by QCanvas** [comment]
bot, build io.github.qcanvas.QCanvasApp

**Comment by QCanvas** [comment]
Build works for me, is this solution acceptable?

**Comment by QCanvas** [comment]
bot, build io.github.qcanvas.QCanvasApp

**Comment by QCanvas** [comment]
> Build [136089 successful](https://buildbot.flathub.org/#/builders/6/builds/136089) To test this build, install it from the testing repository:
> 
> ```
> flatpak install --user https://dl.flathub.org/build-repo/119126/io.github.qcanvas.QCanvasApp.flatpakref
> ```

works

**Comment by QCanvas** [comment]
> Why are you using arch specific .whl packages for everything when sdists exist?

I am using arch specific packages because the sdist build fails for some of the packages. Is there any disadvantage to doing it this way?

**Comment by bbhtt** [comment]
Using sdist is better, because the .whls aren't built against the ABI of the Flatpak runtimes. If it works then it's ok.

**Pull request closed without merging**